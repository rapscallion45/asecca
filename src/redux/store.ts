import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
