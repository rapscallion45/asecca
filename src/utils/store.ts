/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
