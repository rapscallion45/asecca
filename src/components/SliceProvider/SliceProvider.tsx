import { FC, ReactNode, createContext, useContext } from 'react';
import { useSelector } from 'react-redux';
import { Slice } from '@reduxjs/toolkit';
import { AppState } from '@/redux/store';

/**
 * Default value for the slice context
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof SliceProvider
 *
 * @constant
 * @type {Slice<Object>}
 */
const SliceContext = createContext<Slice>({} as Slice);

/**
 * Slice Provider Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef ISliceProviderProps
 * @prop {Slice} slice - context slice to provide down to child nodes
 * @prop {ReactNode} children - child nodes
 */
interface ISliceProviderProps {
  slice: Slice;
  children: ReactNode;
}

/**
 * Slice Provider
 *
 * Slice provider functional component wrapper for providing global state
 * managment functionality to child components
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @component
 * @param {ISliceProviderProps} props - component props
 * @returns {FC} - slice provider functional component
 */
const SliceProvider: FC<ISliceProviderProps> = (props) => {
  const { slice, children } = props;
  return (
    <SliceContext.Provider value={slice}>{children}</SliceContext.Provider>
  );
};

/**
 * Provider specific actions interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {CaseReducerActions<any, string>}
 */
export const useSliceActions = () => useContext<Slice>(SliceContext).actions;

/**
 * Provider specific selector interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {AppState}
 */
export const useSliceSelector = () => {
  const { name } = useContext<Slice>(SliceContext);
  return useSelector((state: AppState) => state[name as keyof AppState]);
};

export default SliceProvider;
