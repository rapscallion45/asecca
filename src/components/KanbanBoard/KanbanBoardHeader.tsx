import { FC } from 'react';
import {
  Box,
  Typography,
  Skeleton,
  Divider,
  LinearProgress,
} from '@mui/material';
import KanbanBoardMenu from '@/components/KanbanBoard/KanbanBoardMenu';
import { useSliceSelector } from '@/components/SliceProvider/SliceProvider';
import { IKanbanBoardState } from '@/redux/types';

/**
 * Kanban Board Header Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.11
 *
 * @typedef IKanbanBoardHeaderProps
 * @prop {string} name - Kanban board name
 */
interface IKanbanBoardHeaderProps {
  name: string;
}

/**
 * Kanban Board Header
 *
 * Application Kanban Board Header interface
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.9
 *
 * @component
 * @returns {FC} - Kanban Board Header functional component
 */
const KanbanBoardHeader: FC<IKanbanBoardHeaderProps> = (props) => {
  const { name } = props;
  const {
    data: boardData,
    error,
    loading,
  } = useSliceSelector() as IKanbanBoardState;

  return (
    <>
      <Box display="flex" pb={1}>
        <Typography variant="h4">
          {!loading && !error ? (
            name
          ) : (
            <Box display="flex">
              <Typography variant="h4">
                {error ?? 'Loading board...'}
              </Typography>
            </Box>
          )}
        </Typography>
        {!loading && !error ? (
          <Box display="flex" justifyContent="end" sx={{ flexGrow: 1 }}>
            <Box display="flex" justifyContent="center" alignItems="center">
              <KanbanBoardMenu currentData={boardData} />
            </Box>
          </Box>
        ) : (
          <Box display="flex" justifyContent="end" sx={{ flexGrow: 1 }}>
            {!error && (
              <>
                <Skeleton
                  variant="rectangular"
                  width={140}
                  height={40}
                  sx={{ borderRadius: '4px' }}
                />
                <Skeleton
                  variant="rectangular"
                  width={70}
                  height={40}
                  sx={{ borderRadius: '4px', ml: 2 }}
                />
              </>
            )}
          </Box>
        )}
      </Box>
      {!loading ? (
        <Divider />
      ) : (
        <LinearProgress color="secondary" sx={{ borderRadius: 1 }} />
      )}
    </>
  );
};

export default KanbanBoardHeader;
