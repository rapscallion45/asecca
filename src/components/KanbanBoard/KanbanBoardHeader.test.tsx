import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, waitFor, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '../../redux/store';
import KanbanBoardHeader from './KanbanBoardHeader';
import SliceProvider from '../SliceProvider/SliceProvider';

/* default test query ID */
const query: string = '66135000001760012';

/* test board name */
const testName: string = 'Test Board';

/**
 * Kanban Board Header Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board Header', () => {
  describe('Values', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render passed board name and data', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardHeader name={testName} />
          </SliceProvider>
        </Provider>
      );

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        expect(screen.getByText(testName)).toBeInTheDocument();
      });
    });
  });

  describe('Error and Loading', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render loading text when loading prop set', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardHeader name={testName} />
          </SliceProvider>
        </Provider>
      );

      /** Assert - loading skeleton and message should be rendered while waiting */
      await waitFor(() => {
        expect(screen.getByText('Loading board...')).toBeInTheDocument();
        expect(
          screen.getByTestId('kanban-board-header-loading-skeleton')
        ).toBeInTheDocument();
        expect(screen.getByRole('progressbar')).toBeInTheDocument();
      });
    });

    // it('Should render error message when error prop set', async () => {
    //   /** Arrange */
    //   const errorMsg: string = 'Failed to load Kanban data from server.';
    //   /** mock kanban services API to return error */
    //   jest.mock('../../services/kanban/collectionsKanbanService', () => ({
    //     getKanbanBoardByProjectId: async () =>
    //       Promise.reject(new Error(errorMsg)),
    //   }));
    //   act(() => {
    //     store.dispatch(
    //       fetchKanbanBoardByProjectId({
    //         projectId: query,
    //       })
    //     );
    //   });

    //   /** Act */
    //   render(
    //     <Provider store={store}>
    //       <SliceProvider slice={collectionsKanbanSlice}>
    //         <KanbanBoardHeader name={testName} />
    //       </SliceProvider>
    //     </Provider>
    //   );

    //   /** Assert - correct value should be displayed from test data */
    //   await waitFor(() => {
    //     expect(screen.getByText(errorMsg)).toBeInTheDocument();
    //   });
    // });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardHeader name={testName} />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
