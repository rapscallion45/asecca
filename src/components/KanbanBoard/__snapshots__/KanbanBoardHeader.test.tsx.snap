// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Header Renders correctly 1`] = `
[
  <div
    className="MuiBox-root css-1c18zwa"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      <div
        className="MuiBox-root css-k008qs"
      >
        <h4
          className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Loading board...
        </h4>
      </div>
    </h4>
    <div
      className="MuiBox-root css-152fgxh"
    >
      <span
        className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-tqb5fd-MuiSkeleton-root"
        data-testid="kanban-board-header-loading-skeleton"
        style={
          {
            "height": 40,
            "width": 140,
          }
        }
      />
      <span
        className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-16y9oml-MuiSkeleton-root"
        style={
          {
            "height": 40,
            "width": 70,
          }
        }
      />
    </div>
  </div>,
  <span
    className="MuiLinearProgress-root MuiLinearProgress-colorSecondary MuiLinearProgress-indeterminate css-v8mbwy-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar1Indeterminate css-p53mz9-MuiLinearProgress-bar1"
      style={{}}
    />
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar2Indeterminate css-xybj4j-MuiLinearProgress-bar2"
      style={{}}
    />
  </span>,
]
`;
